website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+We describe more details for #1 and #2 related to checkpointing optimization as follows.
+
+**Google Cloud Storage and Checkpointing**
+
+GCS serves as our checkpoint store due to its cost-effectiveness, scalability, and durability. When working with GCS, several design considerations and best practices can optimize scaling and performance:
+
+
+
+* Data partitioning methods like range partitioning, which divides data based on specific attributes, and hash partitioning, distributing data evenly using hash functions, are crucial.
+* Avoiding sequential key names, especially timestamps, helps prevent hotspots and uneven data distribution; instead, introduce random prefixes for object distribution.
+* Employing a hierarchical folder structure enhances data management and reduces the number of objects in a single directory.
+* For small files, combining them into larger ones improves read throughput, while minimizing small files reduces inefficient storage use and metadata operations.
+
+**Tune the frequency writing the GCS**
+
+One of our primary challenges lay in scaling jobs efficiently. Stateless jobs, which are relatively simpler, still presented hurdles, especially in scenarios where Flink needed to process an overwhelming number of workers. To overcome this challenge, We increased state.storage.fs.memory-threshold settings to 1 MB from 20KB (??). This configuration allowed us to combine small checkpoints files into larger ones at Job Manager level and reduce metadata calls.
+
+Another challenge we tackled was optimizing the performance of GCS operations. While GCS is excellent for streaming large amounts of data, it does have limitations when it comes to handling high-frequency I/O requests. To mitigate this, we employed strategies like introducing random prefixes in key names, avoiding sequential key names, and optimizing our GCS sharding techniques. These methods significantly enhanced our GCS performance, enabling smooth operation of our stateless jobs.
+
+This is the chart showing the GCS writes reduction after changing the memory-threshold:
+
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/gcs-write-graph.png"
+alt="GCS write Graph">
+
+
+
+# Conclusion
+
+Cortex Data Lake is fully migrated from Dataflow streaming engine to Flink self managed streaming engine infrastructure. We have achieved our goals to run the system more cost efficiently (more than half cost cut), and run the infrastructure on multiple clouds such as GCP and AWS. We have learned how to build a large scale reliable production system based on open sources. We see large potentials to customize the system based on our specific needs as we have a lot of freedom to customize the open source code and configuration. In the next Part 2 post we will give more details on autoscaling and performance tuning parts. We hope our experience will be helpful for readers who will explore similar solutions for their own organizations.
```

Comment:
nit: Maybe rephrase this to "Palo Alto Networks® Cortex Data Lake" since the audience might not know about this product at all and it jumps out of nowhere when the introduction is about PANW.

Reply:
@KevinGG I'll go ahead and resolve this. Please feel free to re-open if you feel it is still unresolved.